executor/parallel_NLJ.go

```diff
@@ -0,0 +1,318 @@
+	// use slices since the inner side is paralleled
+	corCols       [][]*expression.CorrelatedColumn
+	innerFilter   []expression.CNFExprs
+	innerExecs    []Executor
+	innerList     []*chunk.List
+	innerChunk    []*chunk.Chunk
+	innerSelected [][]bool
+	innerIter     []chunk.Iterator
+	outerRow      []*chunk.Row
+	hasMatch      []bool
+	hasNull       []bool
+
+	// fields are used to do concurrency control
+	concurrency int
+	started     bool
+	freeChkCh   chan *chunk.Chunk
+	resultChkCh chan result
+	outerRowCh  chan outerRow
+	numWorkers  int32
+	exit        chan struct{}
+	wg          sync.WaitGroup
+
+	joiner     joiner          // it's stateless
+	memTracker *memory.Tracker // track memory usage.
+}
+
+// Open implements the Executor interface.
+func (e *ParallelNestedLoopApplyExec) Open(ctx context.Context) error {
+	err := e.outerExec.Open(ctx)
+	if err != nil {
+		return err
+	}
+	e.memTracker = memory.NewTracker(e.id, -1)
+	e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
+
+	e.outerList = chunk.NewList(retTypes(e.outerExec), e.initCap, e.maxChunkSize)
+	e.outerList.GetMemTracker().SetLabel(stringutil.StringerStr("outerList"))
+	e.outerList.GetMemTracker().AttachTo(e.memTracker)
+
+	e.innerList = make([]*chunk.List, e.concurrency)
+	e.innerChunk = make([]*chunk.Chunk, e.concurrency)
+	e.innerSelected = make([][]bool, e.concurrency)
+	e.innerIter = make([]chunk.Iterator, e.concurrency)
+	e.outerRow = make([]*chunk.Row, e.concurrency)
+	e.hasMatch = make([]bool, e.concurrency)
+	e.hasNull = make([]bool, e.concurrency)
+	for i := 0; i < e.concurrency; i++ {
+		e.innerChunk[i] = newFirstChunk(e.innerExecs[i])
+		e.innerList[i] = chunk.NewList(retTypes(e.innerExecs[i]), e.initCap, e.maxChunkSize)
+		e.innerList[i].GetMemTracker().SetLabel(innerListLabel)
+		e.innerList[i].GetMemTracker().AttachTo(e.memTracker)
+	}
+
+	e.freeChkCh = make(chan *chunk.Chunk, e.concurrency)
+	e.resultChkCh = make(chan result, e.concurrency+1) // innerWorkers + outerWorker
+	e.outerRowCh = make(chan outerRow)
+	e.exit = make(chan struct{})
+	for i := 0; i < e.concurrency; i++ {
+		e.freeChkCh <- newFirstChunk(e)
+	}
+	return nil
+}
+
+// Next implements the Executor interface.
+func (e *ParallelNestedLoopApplyExec) Next(ctx context.Context, req *chunk.Chunk) (err error) {
+	if !e.started {
+		e.wg.Add(1)
+		go e.outerWorker(ctx)
+		for i := 0; i < e.concurrency; i++ {
+			e.wg.Add(1)
+			atomic.AddInt32(&e.numWorkers, 1)
+			go e.innerWorker(ctx, i)
+		}
+		e.started = true
+	}
+	result := <-e.resultChkCh
+	if result.err != nil {
+		return result.err
+	}
+	req.SwapColumns(result.chk)
+	e.freeChkCh <- result.chk
+	return nil
+}
+
+// Close implements the Executor interface.
+func (e *ParallelNestedLoopApplyExec) Close() error {
+	e.memTracker = nil
+	err := e.outerExec.Close()
+	if e.started {
+		close(e.exit)
+		e.started = false
+		e.wg.Wait()
+	}
+	return err
+}
+
+func (e *ParallelNestedLoopApplyExec) outerWorker(ctx context.Context) {
```

Comment:
Would it break the order property of the output of the outer.

Reply:
You are right. I will add some code to let the planner to determine if the parallel apply can be used.

Reply:
Now I introduce a new step(the function `enableParallelApply()`) in the post-optimization phase to let the planner make the decision. PTAL @fzhedu

Reply:
ok